### PR TITLE
Close cr-text challenge dialog on state change

### PR DESCRIPTION
**Pull request by selvanair**
- All windows are notified when OpenVPN state changes -- cr-text dialog is made to close on receiving this (see issue #440)
- Change the way "challenges" that require no response are processed -- required to handle state-change notification
- Fix a bug in the parsing of the challenge text (we were truncating strings that contain colons)

### CODE REVIEW

**Comment by lstipakov** [comment]
Hi,

After this change CR dialog disappears quickly after popping up. See the difference below (second gif shows current behavior before this patch).

dialog_disappears
dialog_ok

**Comment by selvanair** [comment]
Thanks for testing.
I was under the impression that when auth is pending no state change message arrives unless it times out or auth somehow completes out-of-band. But it looks like state change to GET_CONFIG happens when the auth is still pending. Will have to make this conditional on certain state changes.

Why is the client moving to GET_CONFIG and start sending PUSH_REQUEST when auth is pending? Is that a feature or a bug in OpenVPN?

**Comment by schwabe** [comment]
Feature, kind of. PUSH_REQUEST is basically the keep alive in this state.

**Comment by TinCanTech** [comment]
> Why is the client moving to GET_CONFIG and start sending PUSH_REQUEST when auth is pending? Is that a feature or a bug in OpenVPN?

_sending PUSH_REQUEST when auth is pending_ is a new OpenVPN 2.5 feature.  More commonly known as deferred-auth.

**Comment by selvanair** [comment]
Thank you both. 

I suppose I can assume that the state will change to RECONNECTING or EXITING if auth doesn't complete or to CONNECTED if it succeeds. If so closing the dialog in those cases should be enough. 

This is essentially to handle cases where the challenge says, e.g.,  "Approve the push notification on your phone or input the code from the authenticator app". When the user opts do the former, he is not to be bothered to click OK on this dialog as well.

**Comment by schwabe** [comment]
yes the change will change to something else when the auth doesn't complete.

**Comment by selvanair** [comment]
force pushed with following changes
(i) The bug fix commit removed -- already merged
(ii) The dialog is now auto-closed only if the state changes to something other than GET_CONFIG.

**Comment by lstipakov** [comment]
I haven't tested auto-close, but tested that dialog is displayed, doesn't disappear on GET_CONFIG and, when response is not required (tested by modifying flags in debugger), response-related fields are hidden.

File: openvpn.c [hunk 1]
```diff
@@ -764,6 +778,20 @@ GenericPassDialogFunc(HWND hwndDlg, UINT msg, WPARAM wParam, LPARAM lParam)
...
+    case WM_OVPN_STATE: /* state change message is received from OpenVPN daemon */
+        /*
+         * AUTH_PENDING immediately transitions to GET_CONFIG until
+         * auth succeeds or connection restarts/aborts.
+         * Close the CR_TEXT dialog if state changes to anything
+         * other than GET_CONFIG. The state is in lParam.
+         */
+        param = (auth_param_t *) GetProp(hwndDlg, cfgProp);
+        if ((param->flags & FLAG_CR_TYPE_CRTEXT)
+            && strcmp((const char*) lParam, "GET_CONFIG"))
```
Comment (lstipakov) [inline openvpn.c:790]: space between `char` and `*`

**Review by lstipakov — APPROVED** [review]

File: openvpn.c (hunk 1 above)
Comment (selvanair) [reply openvpn.c]: fixed

**Pull request merged**